### PR TITLE
hinted handoff: improve segment replay logic (superseded by other PR)

### PR DESCRIPTION
**Pull request by piodul**
When sending hints from one file, rps_set is used to keep track of
positions of hints that are currently sent. If sending of a hint fails,
its position is not removed from rps_set. If some hints fail to be sent
while handling a hints file, the lowest position from rps_set is used
to calculate the position from where to start when sending of the file
is retried.

Keeping track of commitlog positions this way isn't necessary to
calculate this position. This patch removes rps_set and replaces it
with first_failed_rp - which is just a single
std::optional<db::replay_position>. This value is updated when a hint
send failure is detected.

This simplifies calculation of starting position for the next retry, and
allowed to remove some error handling logic related to an edge case when
inserting to rps_set fails.

- unit(dev)
- dtest(hintedhandoff_additional_test, dev)

### CODE REVIEW

**Comment by vladzcloudius** [comment]
@piodul You missed one important failure case, which is an error that comes from the commitlog framework. Namely the error that would happen outside the scope of send_one_hint() call and it may happen in the middle of sending hints from one file, e.g. failure of allocation in the commitlog::read_file().

That's why this logic is exactly as it is - we have to keep the current *progress*.

In that case your code for instance would have to re-send the whole segment.

Now let's look at the actual issues you named in the patch description and see how they may be fixed.

The fact that I don't remove the `rp` from the `rp_set` in the hint discard case is obviously a bug and it's a single line fix which needs to be done.

The size of the `rp_set` is meant to be limited by maximum parallelism of hints sending flow - as is written in the field's description.

If memory serves me well this value is 128 today and if we want to minimize re-hash events we may pre-allocate the corresponding buckets count - there is a corresponding constructor.

If we want to avoid it completely on account of some extra CPU (if at all) we may want to use a different data structure, e.g. intrusive set or list or event a vector.

**Comment by piodul** [comment]
> @piodul You missed one important failure case, which is an error that comes from the commitlog framework. Namely the error that would happen outside the scope of send_one_hint() call and it may happen in the middle of sending hints from one file, e.g. failure of allocation in the commitlog::read_file().

It's true that I missed this case, but I don't think that current logic handles it.

"rps_set" only holds replay positions of hints that are either being sent at the moment, or failed to be sent. If an error is thrown inside commitlog::read_log_file, we will still wait for all in-progress send_one_hint operations to complete (which is done [here](https://github.com/scylladb/scylla/blob/7da949026da980caf1cee0d96cce1dc76e6e736c/db/hints/manager.cc#L771)). If all of them complete successfully, then rps_set will be empty and we won't advance _last_not_complete_rp.

I made another patch, dependent on this one, that improves retry behavior in this case - we'll keep track of the last replay position that commitlog::read_log_file's callback was called with, and we'll include it when calculating the restart position (equivalent to min(rps_set ∪ {last_replay_rp}). It's not perfect - if an error is thrown in commitlog::send_one_file, we will have to re-sent last processed hint, but it is much better than retrying from the previous, not updated "_last_not_complete_rp".

Link to the new patch: https://github.com/scylladb/scylla/pull/6416

> The fact that I don't remove the rp from the rp_set in the hint discard case is obviously a bug and it's a single line fix which needs to be done.

This patch started as a one-liner to fix this exact issue, but then I realized we could remove rps_set and remove the need to handle the case in which inserting to it fails.

**Comment by psarna** [comment]
@vladzcloudius I think that this patch is a net win - instead of maintaining a whole ordered data structure, the bookkeeping is reduced to a single variable, and the commitlog problem is handled separately in #6416 .
Still, perhaps it makes sense to split this original small fix (removing the hint from the set when it's discarded) as a small separate commit, since it will be easy to backport it to other releases if need be.

**Comment by piodul** [comment]
@psarna I uploaded the fix here: https://github.com/scylladb/scylla/pull/6417

I'll update description of this patch shortly, so that it won't mention the issue with old hints.

**Comment by piodul** [comment]
v2: updated commit and PR description so that they don't mention the issue with old hints being kept in rps_set - this is handled in #6417

File: db/hints/manager.cc [hunk 1]
```diff
@@ -724,15 +714,25 @@ future<> manager::end_point_hints_manager::sender::send_one_hint(lw_shared_ptr<s
             } catch (no_column_mapping& e) {
                 manager_logger.debug("send_hints(): {} at {}: {}", fname, rp, e.what());
                 ++this->shard_stats().discarded;
+            } catch (...) {
+                manager_logger.debug("send_hints(): unexpected error in file {} at {}: {}", fname, rp, std::current_exception());
+                ctx_ptr->on_hint_send_failure(rp);
```
Comment (haaawk) [inline db/hints/manager.cc:719]: Should we rethrow the exception here?

File: db/hints/manager.cc [hunk 2]
```diff
@@ -778,11 +778,8 @@ bool manager::end_point_hints_manager::sender::send_one_file(const sstring& fnam
 
     // update the next iteration replay position if needed
     if (ctx_ptr->state.contains(send_state::segment_replay_failed)) {
-        if (ctx_ptr->state.contains(send_state::restart_segment)) {
```
Comment (haaawk) [inline db/hints/manager.cc:781]: Why is it safe to remove this case?

File: db/hints/manager.hh [hunk 3]
```diff
@@ -97,21 +98,21 @@ public:
 
             enum class send_state {
                 segment_replay_failed,  // current segment sending failed
```
Comment (haaawk) [inline db/hints/manager.hh:100]: Since there's only one state now, can we change the field that has type send_state_set to bool?

**Comment by vladzcloudius** [comment]
> > @piodul You missed one important failure case, which is an error that comes from the commitlog framework. Namely the error that would happen outside the scope of send_one_hint() call and it may happen in the middle of sending hints from one file, e.g. failure of allocation in the commitlog::read_file().
> 
> It's true that I missed this case, but I don't think that current logic handles it.
> 
> "rps_set" only holds replay positions of hints that are either being sent at the moment, or failed to be sent. If an error is thrown inside commitlog::read_log_file, we will still wait for all in-progress send_one_hint operations to complete (which is done [here](https://github.com/scylladb/scylla/blob/7da949026da980caf1cee0d96cce1dc76e6e736c/db/hints/manager.cc#L771)). If all of them complete successfully, then rps_set will be empty and we won't advance _last_not_complete_rp.

@piodul I think you might have missed that `commitlog::read_log_file()` updates the `_last_not_complete_rp.pos`, ifn case `rps_set` becomes empty after the gate this is exactly where you want to start sending from the next time.

> 
> I made another patch, dependent on this one, that improves retry behavior in this case - we'll keep track of the last replay position that commitlog::read_log_file's callback was called with, and we'll include it when calculating the restart position (equivalent to min(rps_set ∪ {last_replay_rp}). It's not perfect - if an error is thrown in commitlog::send_one_file, we will have to re-sent last processed hint, but it is much better than retrying from the previous, not updated "_last_not_complete_rp".

You don't need to track that - read_log_file() already does it for you. The problem is that you don't know which mutation failed, namely knowing the last doesn't help you much because some segments before it could have failed.

That's why the code is exactly as it is.
Like I said, I don't think it requires improving except for that single liner that would remove the hint from the set when it has aged.

> 
> Link to the new patch: #6416
> 
> > The fact that I don't remove the rp from the rp_set in the hint discard case is obviously a bug and it's a single line fix which needs to be done.
> 
> This patch started as a one-liner to fix this exact issue, but then I realized we could remove rps_set and remove the need to handle the case in which inserting to it fails.

**Comment by vladzcloudius** [comment]
> @vladzcloudius I think that this patch is a net win - instead of maintaining a whole ordered data structure, the bookkeeping is reduced to a single variable, and the commitlog problem is handled separately in #6416 .
> Still, perhaps it makes sense to split this original small fix (removing the hint from the set when it's discarded) as a small separate commit, since it will be easy to backport it to other releases if need be.

@psarna There isn't any ordered data structure in this context: `rps_set` is an unordered_set, which is a hash table.
I don't think this series is a net win. It introduces bugs like I described before.
More than that I don't see any issue with the current code except for a tiny bug I mentioned above. 

I don't think that a dynamic hash table resizing would present any issue if occurs - this use case is already addressed in the present code here: https://github.com/scylladb/scylla/blob/7da949026da980caf1cee0d96cce1dc76e6e736c/db/hints/manager.cc#L781

I think this series is trying to fix something that ain't broken.

**Comment by vladzcloudius** [comment]
BTW, there is another place where the `rp` may "leak" in the current code and it's when you get to this `return`: https://github.com/scylladb/scylla/blob/7da949026da980caf1cee0d96cce1dc76e6e736c/db/hints/manager.cc#L728

So, another single liner similar to https://github.com/scylladb/scylla/pull/6417

**Comment by piodul** [comment]
@haaawk @vladzcloudius Because it looks like the same changes are reviewed both there and in PR https://github.com/scylladb/scylla/pull/6416, and Vlad suggested to keep everything in a single patch, I'd like to move discussion to the PR linked in this comment. I'll publish further updates there. I'll close this PR shortly.

File: db/hints/manager.cc (hunk 1 above)
Comment (piodul) [reply db/hints/manager.cc:719]: Done. Addressed in https://github.com/scylladb/scylla/pull/6416

File: db/hints/manager.cc (hunk 2 above)
Comment (piodul) [reply db/hints/manager.cc:781]: We used to set this flag only when inserting to rps_set failed. Now, this flag is never set, so we can remove this case.

File: db/hints/manager.hh (hunk 3 above)
Comment (piodul) [reply db/hints/manager.hh:100]: Done, addressed in https://github.com/scylladb/scylla/pull/6416

**Comment by piodul** [comment]
@vladzcloudius 

> BTW, there is another place where the rp may "leak" in the current code and it's when you get to this return:

Good catch, I'll publish a patch for that shortly.

**Pull request closed without merging**